### ISSUE TITLE
SFTTrainer does not support a custom metric for evaluation (compute_metrics)

### ISSUE THREAD

**Issue by eranhirs** [huggingface/trl#1222]
# Desired behavior
As far as I can tell, `SFTTrainer` does not support a custom metric for evaluation (`compute_metrics`). This makes it hard to track the fine-tuning of a model, implement an early stopping method, or interchange with the `Seq2SeqTrainer` if we want to try fine-tuning models with different architectures.

# Details

Since `trl` is designed for generation models, the `SFTTrainer` should also call `.generate()` in its `prediction_step`. This is similar to the `Seq2SeqTrainer`  ([link](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer_seq2seq.py#L232C17-L232C17)).

In addition, the `input_ids` used for generation should not contain the labels when calling `.generate()`. Possibly, a new data collator needs to be created, one which is similar to `DataCollatorForSeq2Seq`. However, since we still want to calculate the eval loss, the new data collator should also contain the functionality of `DataCollatorForCompletionOnlyLM`. Also, this requires `SFTTrainer` to receive a separate data collator to use for the eval dataset from the training dataset.

I would appreciate your inputs on this. If this makes sense, I can create a PR and provide an example in the `sft.py` script.

**Comment by younesbelkada** [comment]
Hi @eranhirs 
Thanks very much for this feature request! As fas as I can tell, this makes sense to me ! Would you be happy to open a PoC PR and we can brainstorm from there about the exact API?

**Comment by harsh-ub** [comment]
Hi! I'm facing a similar issue with generation where calling a `generate()` method with the model in `trl`'s `SFTTrainer` doesn't work out of box. Any thoughts how we can use the generate during training (eg. via a callback).

**Comment by nai-kon** [comment]
Hi, I have a similar issue with this. Is there any plans to respond?

**Comment by nai-kon** [comment]
Also, could you please stop closing the issue just because there hasn't been any interaction for a while even though it hasn't been resolved?
This makes it difficult to accurately determine whether the issue has been resolved and closed.

**Comment by kaiguy23** [comment]
I would also be interested in this feature. I assume it has not been solved yet.

**Comment by ytfksw** [comment]
I am interested in this feature and would like to consider creating a pull request.

**Comment by haryoa** [comment]
need this feature.

**Comment by yznlp** [comment]
Not sure if someone is working on this, but for what it's worth it seems like you can get pretty far by subclassing `trl.SFTTrainer` and `transformers.Seq2SeqTrainer` since they both inherit from the `Trainer` class. Something like

Get new config class so you can pass `generation_config` into training args
```python
@dataclass
class SFTSeq2SeqConfig(SFTConfig, Seq2SeqTrainingArguments):
    pass
```

init mostly unchanged, just need to parse `args.generation_config`
```python
class SFTSeq2SeqTrainer(SFTTrainer, Seq2SeqTrainer):
    def __init__(self, **kwargs):
        SFTTrainer.__init__(self, **kwargs)
        # relevant bit from Seq2SeqTrainer.__init__()
        gen_config = Seq2SeqTrainer.load_generation_config(self.args.generation_config)
        self.model.generation_config = gen_config
```

the main method to override is `prediction_step()` and I left it mostly unchanged from `SFTTrainer.prediction_step()` because it's mostly handling setups and integrations. 
```python
    def prediction_step(..., **gen_kwargs):
        ### setups
        
        generated_tokens = self.model.generate(..., **gen_kwargs)

        ### loss calculations if you need
```

then to get different behaviours from data collator between train and evaluate you can override `get_train_dataloader()` and `get_eval_dataloader()`. I just added corresponding `collate_train()` and `collate_eval()` to `MyDataCollator`. 

```python
    def get_train_dataloader(self):
        # same as `Trainer.get_train_dataloader()` except the one change

        # old 
        data_collator = self.data_collator
        # new
        data_collator = self.data_collator.collate_train
```

A bit messy but hope this helps

**Comment by BenBatsir** [comment]
@yznlp Any chance you can share the full example?

**Comment by eranhirs** [comment]
Just run into [this Wandb tutorial](https://wandb.ai/capecape/alpaca_ft/reports/How-to-Fine-tune-an-LLM-Part-3-The-HuggingFace-Trainer--Vmlldzo1OTEyNjMy#(optional)-preprocessing:-masking-instructions-by-using-the-datacollatorforcompletiononlylm), it suggests a nice workaround of using a callback during training. The disadvantage of this approach is that the eval dataset will be evaluated twice: once to calculate the eval loss, inside `SFTTrainer` evaluation loop, and once inside the callback.

**Comment by yznlp** [comment]
> @yznlp Any chance you can share the full example?

Sorry for the delay. What I have is quite rough around the edges but does the job for me - I'm not outputting the eval loss for example. I got rid of some hard coded bits for my model so there might be some minor errors. 

Tested with `transformers==4.44.2 trl==0.9.6 datasets==2.9.0 accelerate==1.0.0`

```python
from typing import Optional, Union, Dict, Tuple, List, Any
from dataclasses import dataclass

import torch
import torch.nn as nn
from torch.utils.data import DataLoader

from trl import (
    SFTTrainer,
    SFTConfig,
)
from transformers import (
    Seq2SeqTrainer, 
    Seq2SeqTrainingArguments,
)
from transformers.trainer_utils import seed_worker
from transformers.trainer_pt_utils import nested_detach
from transformers.utils import is_sagemaker_mp_enabled, is_datasets_available
from transformers.integrations.deepspeed import is_deepspeed_zero3_enabled
import datasets

from accelerate import Accelerator
accelerator = Accelerator()

@dataclass
class SFTSeq2SeqConfig(SFTConfig, Seq2SeqTrainingArguments):
    pass

class SFTSeq2SeqTrainer(SFTTrainer, Seq2SeqTrainer):
    def __init__(self, **kwargs):
        SFTTrainer.__init__(self, **kwargs)
        # relevant bit from Seq2SeqTrainer.__init__()
        gen_config = Seq2SeqTrainer.load_generation_config(self.args.generation_config)
        self.model.generation_config = gen_config

    def get_train_dataloader(self) -> DataLoader:
        # same as Trainer.get_tarin_dataloader() except that one bit
        if self.train_dataset is None:
            raise ValueError("Trainer: training requires a train_dataset.")

        train_dataset = self.train_dataset
        # only change
        data_collator = self.data_collator.collate_train
        if is_datasets_available() and isinstance(train_dataset, datasets.Dataset):
            train_dataset = self._remove_unused_columns(train_dataset, description="training")
        else:
            data_collator = self._get_collator_with_removed_columns(data_collator, description="training")

        dataloader_params = {
            "batch_size": self._train_batch_size,
            "collate_fn": data_collator,
            "num_workers": self.args.dataloader_num_workers,
            "pin_memory": self.args.dataloader_pin_memory,
            "persistent_workers": self.args.dataloader_persistent_workers,
        }

        if not isinstance(train_dataset, torch.utils.data.IterableDataset):
            dataloader_params["sampler"] = self._get_train_sampler()
            dataloader_params["drop_last"] = self.args.dataloader_drop_last
            dataloader_params["worker_init_fn"] = seed_worker
            dataloader_params["prefetch_factor"] = self.args.dataloader_prefetch_factor

        return self.accelerator.prepare(DataLoader(train_dataset, **dataloader_params))

    def get_eval_dataloader(self, eval_dataset: Optional[Union[str, datasets.Dataset]] = None) -> DataLoader:
        if eval_dataset is None and self.eval_dataset is None:
            raise ValueError("Trainer: evaluation requires an eval_dataset.")

        # If we have persistent workers, don't do a fork bomb especially as eval datasets
        # don't change during training
        dataloader_key = eval_dataset if isinstance(eval_dataset, str) else "eval"
        if (
            hasattr(self, "_eval_dataloaders")
            and dataloader_key in self._eval_dataloaders
            and self.args.dataloader_persistent_workers
        ):
            return self.accelerator.prepare(self._eval_dataloaders[dataloader_key])

        eval_dataset = (
            self.eval_dataset[eval_dataset]
            if isinstance(eval_dataset, str)
            else eval_dataset
            if eval_dataset is not None
            else self.eval_dataset
        )
        # only change
        data_collator = self.data_collator.collate_eval

        if is_datasets_available() and isinstance(eval_dataset, datasets.Dataset):
            eval_dataset = self._remove_unused_columns(eval_dataset, description="evaluation")
        else:
            data_collator = self._get_collator_with_removed_columns(data_collator, description="evaluation")

        dataloader_params = {
            "batch_size": self.args.eval_batch_size,
            "collate_fn": data_collator,
            "num_workers": self.args.dataloader_num_workers,
            "pin_memory": self.args.dataloader_pin_memory,
            "persistent_workers": self.args.dataloader_persistent_workers,
        }

        if not isinstance(eval_dataset, torch.utils.data.IterableDataset):
            dataloader_params["sampler"] = self._get_eval_sampler(eval_dataset)
            dataloader_params["drop_last"] = self.args.dataloader_drop_last
            dataloader_params["prefetch_factor"] = self.args.dataloader_prefetch_factor

        # accelerator.free_memory() will destroy the references, so
        # we need to store the non-prepared version
        eval_dataloader = DataLoader(eval_dataset, **dataloader_params)
        if self.args.dataloader_persistent_workers:
            if hasattr(self, "_eval_dataloaders"):
                self._eval_dataloaders[dataloader_key] = eval_dataloader
            else:
                self._eval_dataloaders = {dataloader_key: eval_dataloader}

        return self.accelerator.prepare(eval_dataloader)

    def train(self, *args, **kwargs):
        return SFTTrainer.train(self, *args, **kwargs)

    def evaluate(self, **kwargs):
        # Seq2SeqTrainer.evaluate() calls Trainer.evaluate()
        # Trainer.evaluate() calls Trainer.evaluation_loop() which calls self.prediction_step()

        return Seq2SeqTrainer.evaluate(self, **kwargs)
    
    def predict(self, **kwargs): 
        # same as self.evaluate() except this goes to Trainer.predict()
        # but ultimately still call self.prediction_step()
        return Seq2SeqTrainer.predict(self, **kwargs)

    # the main method that requires overriding
    def prediction_step(
        self,
        model: nn.Module,
        inputs: Dict[str, Union[torch.Tensor, Any]],
        prediction_loss_only: bool,
        ignore_keys: Optional[List[str]] = None,
        calculate_loss: Optional[bool] = False,
        **gen_kwargs,
    ) -> Tuple[Optional[torch.Tensor], Optional[torch.Tensor], Optional[torch.Tensor]]:
        """
        Perform an evaluation step on `model` using `inputs`.

        Subclass and override to inject custom behavior.

        Args:
            model (`nn.Module`):
                The model to evaluate.
            inputs (`Dict[str, Union[torch.Tensor, Any]]`):
                The inputs and targets of the model.

                The dictionary will be unpacked before being fed to the model. Most models expect the targets under the
                argument `labels`. Check your model's documentation for all accepted arguments.
            prediction_loss_only (`bool`):
                Whether or not to return the loss only.
            ignore_keys (`List[str]`, *optional*):
                A list of keys in the output of your model (if it is a dictionary) that should be ignored when
                gathering predictions.

        Return:
            Tuple[Optional[torch.Tensor], Optional[torch.Tensor], Optional[torch.Tensor]]: A tuple with the loss,
            logits/tokens and labels (each being optional).
        """

        # lifted from Seq2SeqTrainer.prediction_step()
        # if we don't want to run prediction using model.generate() then default implementation will do
        if not self.args.predict_with_generate or prediction_loss_only:
            return SFTTrainer.prediction_step(
                self, model, inputs, prediction_loss_only = prediction_loss_only, ignore_keys = ignore_keys
            )

        # lifted from SFTTrainer.prediction_step() - looks like mostly setup stuff
        has_labels = False if len(self.label_names) == 0 else all(inputs.get(k) is not None for k in self.label_names)
        # For CLIP-like models capable of returning loss values.
        # If `return_loss` is not specified or being `None` in `inputs`, we check if the default value of `return_loss`
        # is `True` in `model.forward`.
        return_loss = inputs.get("return_loss", None)
        if return_loss is None:
            return_loss = self.can_return_loss
        loss_without_labels = True if len(self.label_names) == 0 and return_loss else False

        inputs = self._prepare_inputs(inputs)
        if ignore_keys is None:
            if hasattr(self.model, "config"):
                ignore_keys = getattr(self.model.config, "keys_to_ignore_at_inference", [])
            else:
                ignore_keys = []

        # labels may be popped when computing the loss (label smoothing for instance) so we grab them first.
        if has_labels or loss_without_labels:
            labels = nested_detach(tuple(inputs.get(name) for name in self.label_names))
            if len(labels) == 1:
                labels = labels[0]
        else:
            labels = None

        # lifted from Seq2SeqTrainer.prediction_step() - setup for gen_kwargs
        # Priority (handled in generate):
        # non-`None` gen_kwargs > model.generation_config > default GenerationConfig()
        if len(gen_kwargs) == 0 and hasattr(self, "_gen_kwargs"):
            gen_kwargs = self._gen_kwargs.copy()
        if "num_beams" in gen_kwargs and gen_kwargs["num_beams"] is None:
            gen_kwargs.pop("num_beams")
        if "max_length" in gen_kwargs and gen_kwargs["max_length"] is None:
            gen_kwargs.pop("max_length")

        default_synced_gpus = True if is_deepspeed_zero3_enabled() else False
        gen_kwargs["synced_gpus"] = (
            gen_kwargs["synced_gpus"] if gen_kwargs.get("synced_gpus") is not None else default_synced_gpus
        )

        # first "forward" pass - generate report as in inference
        generated_tokens = self.model.generate(**inputs, **gen_kwargs)

        if not calculate_loss:
            return None, generated_tokens, labels
        
        # calculate loss with second forward pass
        # do we need to add label smoothing?
        with torch.no_grad():
            if is_sagemaker_mp_enabled():
                # i don't think we're ever here
                raise NotImplementedError('prediction_step() not implemented for is_sagemaker_mp_enabled() is True')
            
            if has_labels or loss_without_labels:
                with self.compute_loss_context_manager():
                    loss, outputs = self.compute_loss(model, inputs, return_outputs=True)
                loss = loss.mean().detach()

                if isinstance(outputs, dict):
                    logits = tuple(v for k, v in outputs.items() if k not in ignore_keys + ["loss"])
                else:
                    logits = outputs[1:]
            else:
                # shouldn't be here either i think
                loss = None
                with self.compute_loss_context_manager():
                    outputs = model(**inputs)
                if isinstance(outputs, dict):
                    logits = tuple(v for k, v in outputs.items() if k not in ignore_keys)
                else:
                    logits = outputs
                # TODO: this needs to be fixed and made cleaner later.
                if self.args.past_index >= 0:
                    self._past = outputs[self.args.past_index - 1]

        if prediction_loss_only:
            # shouldn't ever be here
            raise ValueError
            return (loss, None, None)

        # carryover from SFTTrainer - we don't need to output logits here
        # logits = nested_detach(logits)
        # if len(logits) == 1:
        #     logits = logits[0]

        return loss, generated_tokens, labels
    


```

**Comment by BenBatsir** [comment]
@yznlp Thanks for sharing.
Sadly I wasn't able to use it

**Comment by steveepreston** [comment]
+1

**Comment by kushwanth22** [comment]
+1

**Comment by kharabyasser** [comment]
+1

**Comment by Outliers1106** [comment]
+1

**Comment by cogrow-tech** [comment]
+1

**Comment by hrshtkpr** [comment]
+1

**Comment by sapirharary** [comment]
+1

**Comment by azhao8** [comment]
+1

**Comment by Axe--** [comment]
+1

**Comment by steveepreston** [comment]
is there any progress on this?
why this is closed?

**Comment by dinusha94** [comment]
Any updates on this? It will be better if we can have something like the following, where we can do a sample inference in the middle

**Callback function**

```
import torch
from transformers import TrainerCallback

class InferenceCallback(TrainerCallback):
    def __init__(self, tokenizer, val_dataset, every_n_steps=100):
        self.tokenizer = tokenizer
        self.val_dataset = val_dataset
        self.every_n_steps = every_n_steps

    def on_step_end(self, args, state, control, **kwargs):
        if state.global_step % self.every_n_steps != 0:
            return

        model = kwargs["model"]
        device = model.device

        model.eval()
        indices = random.sample(range(len(self.val_dataset)), k=1)

        for sample in self.val_dataset.select(indices):  # limit for speed
            prompt = self.tokenizer.apply_chat_template(
                sample["messages"],
                tokenize=False,
                add_generation_prompt=True
            )

            kk = prompt.split('<start_of_turn>model')
            input_prompt = f"{kk[0]}<start_of_turn>model"
            target_response = kk[1].split('<end_of_turn>')[0]

            target_token_len = len(
                            self.tokenizer(
                                target_response,
                                add_special_tokens=False,
                                return_attention_mask=False,
                                return_token_type_ids=False,
                            )["input_ids"]
                        )

            inputs = self.tokenizer(input_prompt, return_tensors="pt").to(device)
            input_len = inputs["input_ids"].shape[-1]

            with torch.no_grad():
                outputs = model.generate(
                    **inputs,
                    max_new_tokens=target_token_len + 16,
                    temperature = 1.0, top_p = 0.95, top_k = 64,
                    # do_sample=True,
                )

            # Slice ONLY generated tokens
            generated_tokens = outputs[0][input_len:]
            generated_response = self.tokenizer.decode(generated_tokens, skip_special_tokens=True)
            log_inference_results(LOG_FILE, MODEL_NAME, generated_response, target_response)


        model.train()

```

**Usage**

```
training_args = SFTConfig(
        dataset_text_field="text",

        per_device_train_batch_size=BATCH_SIZE,
        gradient_accumulation_steps=GRAD_ACC,
        num_train_epochs = 1, # Set this for 1 full training run.
        # max_steps = 100,

        learning_rate=LR,
        warmup_steps=30,

        logging_steps=1,
        optim="adamw_8bit",
        lr_scheduler_type="linear",
        weight_decay=0.001,

        seed=SEED,

        output_dir=OUTPUT_DIR,
        save_strategy="steps",
        save_steps=STEP_CHUNK,
        save_total_limit=2,

        report_to="tensorboard",
    )

trainer = SFTTrainer(
  model=model,
  tokenizer=tokenizer, # Re-added this line
  train_dataset=train_dataset,
  eval_dataset=None,
  args=training_args,
  callbacks=[
        InferenceCallback(
            tokenizer=tokenizer,
            val_dataset=val_dataset,
            every_n_steps=50
        )
    ]
)

```

**Comment by ngocuyen1207** [comment]
Please reopen this

**Comment by qgallouedec** [comment]
I think it's solved:

```python
from datasets import load_dataset
from trl import SFTConfig, SFTTrainer

dataset = load_dataset("trl-internal-testing/zen", "standard_language_modeling")

def dummy_compute_metrics(eval_pred):
    return {"my_metric": 0.123}

training_args = SFTConfig(
    eval_strategy="steps",
    eval_steps=3,
)

trainer = SFTTrainer(
    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    args=training_args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["test"],
    compute_metrics=dummy_compute_metrics,
)

trainer.train()
```

**Comment by steveepreston** [comment]
@qgallouedec Need Confirmation. Can any full test this?
